README
NAME
    Module::Build::Bundle - subclass for supporting Tasks and Bundles

VERSION
    This documentation describes version 0.13

SYNOPSIS
        #In your Build.PL
        use Module::Build::Bundle;

        #Example lifted from: Perl::Critic::logicLAB
        my $build = Module::Build::Bundle->new(
            dist_author   => 'Jonas B. Nielsen (jonasbn), <jonasbn@cpan.org>',
            module_name   => 'Perl::Critic::logicLAB',
            license       => 'artistic',
            create_readme => 1,
            requires      => {
                'Perl::Critic::Policy::logicLAB::ProhibitUseLib' => '0',
                'Perl::Critic::Policy::logicLAB::RequireVersionFormat' => '0',
            },
        );

        $build->create_build_script();


        #In your shell
        % ./Build contents

        #Or implicitly executing contents action
        % ./Build

DESCRIPTION
  FEATURES
    *   Autogeneration of POD for Bundle and Task distributions via a build
        action

    *   Links to required/listed distributions, with or without versions

    *   Links to specific versions of distributions for perl 5.12.0 or newer
        if a version is specified

    *   Inserts a POD section named CONTENTS or something specified by the
        caller

    This module adds a very basic action for propagating a requirements list
    from a Build.PL file's requires section to the a POD section in a
    designated distribution.

SUBROUTINES/METHODS
  ACTION_contents
    This is the build action parsing the requirements specified in the
    Build.PL file. It creates a POD section (see also "FEATURES" above).

    By default it overwrites the CONTENTS section with a POD link listing.
    You can specify a note indicating if what section you want to overwrite
    using the section_header note.

        #Example lifted from: Perl::Critic::logicLAB
        my $build = Module::Build::Bundle->new(
            dist_author   => 'Jonas B. Nielsen (jonasbn), <jonasbn@cpan.org>',
            module_name   => 'Perl::Critic::logicLAB',
            license       => 'artistic',
            create_readme => 1,
            requires      => {
                'Perl::Critic::Policy::logicLAB::ProhibitUseLib' => '0',
                'Perl::Critic::Policy::logicLAB::RequireVersionFormat' => '0',
            },
        );

        $build->notes('section_header' => 'POLICIES');

        $build->create_build_script();

    The section of course has to be present.

    Based on your version of perl and your Build.PL requirements, the links
    will be rendered in the following formats:

    Basic:

        #Build.PL
        requires => {
            'Some::Package' => '0',
        }

        #POD, perl all
        =item * L<Some::Package|Some::Package>

    With version:

        #Build.PL
        requires => {
            'Some::Package' => '1.99',
        }

        #POD, perl < 5.12.0
        =item * L<Some::Package|Some::Package>, 1.99

        #POD, perl >= 5.12.0
        =item * L<Some::Package|Some::Package>, L<1\.99\|http://search.cpan.org/dist/Some-Package-1.99/lib/Some/Package.pm>

  ACTION_build
    This is a simple wrapper around the standard action: Module::Build build
    action. It checks whether "ACTION_contents" have been executed, if not
    it executes it.

  create_mymeta
    This method has been lifted from Module::Build::Base and altered.

    It sets the:

    *   'generated by <package> version <package version>' string in
        MYMETA.yml

    For Module::Build::Bundle:

        #Example MYMETA.yml
        configure_requires:
            Module::Build::Bundle: 0.01
        generated_by: 'Module::Build::Bundle version 0.01'

  get_metadata
    This method has been lifted from Module::Build::Base and altered.

    It sets:

    *   'generated by <package> version <package version>' string in
        META.yml

    *   configure_requires: <package>: <version>

    For Module::Build::Bundle:

        #Example META.yml
        configure_requires:
            Module::Build::Bundle: 0.01
        generated_by: 'Module::Build::Bundle version 0.01'

  do_create_metafile
    This method has been lifted from Module::Build::Base and altered.

    The method was overwritten to be more testable. The method created the
    relevant META file.

DIAGNOSTICS
    *   No <section> section to be replaced

        If the POD to be updated does not contain a placeholder section the
        action will die with the above message.

        The default minimal section should look something like:

            =head1 CONTENTS

            =head1

        Or if you provide your own section_header

            =head1 <section header>

            =head1

CONFIGURATION AND ENVIRONMENT
  CONTENTS SECTION
    The module does per default look for the section named: CONTENTS.

    This is the section used in Bundles, this can be overwritten using the
    section parameter.

    For example Perl::Critic::logicLAB uses a section named POLICIES and
    Task::BeLike::JONASBN uses DEPENDENCIES.

    The problem is that the section has to be present or else the contents
    action will throw an error.

    Module::Build::Bundle is primarily aimed at Bundle distributions. Their
    use is however no longer recommended and Task provides a better way.

DEPENDENCIES
    *   perl 5.6.0

    *   Module::Build::Base, part of the Module::Build distribution

INCOMPATIBILITIES
    The distribution requires perl version from 5.6.0 and up.

BUGS AND LIMITATIONS
    Currently Module::Build::Bundle is not able to handle root based
    distributions meaning distributions with a single Perl module located in
    the root directory instead of the lib structure.

    Apart from that there are no known special limitations or bugs at this
    time, but I am certain there are plenty of scenarios is distribution
    packaging the module is not currently handling.

    The module only supports Bundle/Task distributions based on
    Module::Build. The implementation is based on a subclass of
    Module::Build, which can replace Module::Build in your Build.PL (See:
    "SYNOPSIS").

    As described previously in the documentation a section of documentation
    can only replaced. A section with the generated contents cannot be added
    with out a placeholder in the form of designated section title. This
    might be changed in the future.

    Before version 0.11 the designated module was worked on in lib/, I am
    still unsure as to what the right place to do this is. Perhaps *hooking*
    into the build phase is not a good idea at all.

BUG REPORTING
    Please report any bugs or feature requests via:

    *   email: bug−module-build-bundle at rt.cpan.org

    *   HTTP:
        <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Module-Build-Bundle>

TEST AND QUALITY
  TEST COVERAGE
        ---------------------------- ------ ------ ------ ------ ------ ------ ------
        File                           stmt   bran   cond    sub    pod   time  total
        ---------------------------- ------ ------ ------ ------ ------ ------ ------
        lib/Module/Build/Bundle.pm     44.5   18.3   23.0   70.5  100.0  100.0   39.6
        Total                          44.5   18.3   23.0   70.5  100.0  100.0   39.6
        ---------------------------- ------ ------ ------ ------ ------ ------ ------

    The above coverage report is based on release 0.13

QUALITY AND CODING STANDARD
    The code passes Perl::Critic tests a severity: 1 (brutal)

    The following policies have been disabled:

           Perl::Critic::Policy::InputOutput::RequireBracedFileHandleWithPrint

    Perl::Critic resource file, can be located in the t/ directory of the
    distribution see F <t/perlcriticrc>

    Perl::Tidy resource file, can be obtained from the following URL:

    *   <https://logiclab.jira.com/wiki/display/OPEN/Perl-Tidy>

DEVELOPMENT
TODO
    Please see:
    <https://logiclab.jira.com/browse/MBB#selectedTab=com.atlassian.jira.plu
    gin.system.project%3Aroadmap-panel>

SEE ALSO
    *   Task

    *   TaskBeLike::JONASBN

    *   Perl::Critic::logicLAB

    *   CPAN

    *   CPAN::Bundle
        <http://cpansearch.perl.org/src/ANDK/CPAN-1.9402/lib/CPAN/Bundle.pm>

    *   <https://logiclab.jira.com/wiki/display/OPEN/Module-Build-Bundle>

MOTIVATION
    The motivation was driven by two things.

    *   The joy of fooling around with Module::Build

    *   The need for automating the documentation generation

    I have a few perks and one of them is that I never get to automate stuff
    until very late and I always regret that. So when I released
    Bundle::JONASBN, now Task::BeLike::JONASBN::Task::BeLike::JONASBN I
    thought I might aswell get it automated right away.

    This module lived for a long time as a part of Bundle::JONASBN but then
    I needed it for some other distributions, so I decided to separate it
    out.

ACKNOWLEDGEMENTS
    *   Adam Kennedy (ADAMK) author of Task, a very basic and simple
        solution

    *   The Module::Build developers

    *   Lars Dɪᴇᴄᴋᴏᴡ (DAXIM) for reporting RT:83754, resulting in release
        0.11

    *   Andreas J. König (ANDK) for reporting RT:82128, included in release
        0.10

AUTHOR
    *   Jonas B. Nielsen (jonasbn) "<jonasbn@cpan.org>"

LICENSE AND COPYRIGHT
    Copyright 2010-2014 jonasbn, all rights reserved.

    Module::Build::Bundle is released under the Artistic License 2.0



Changes
Changes file for Module::Build::Bundle

HEAD

- Added is_windowsish as required by Module::Build for sub-classing


0.16 2015-09-06 Bug fix release, update not required

- Fixed issue with test dependencies, Text::Soundex was missing

- Addressed issue #2, File::stat used instead of File::Stat

  https://github.com/jonasbn/mbb/issues/2
  https://rt.cpan.org/Ticket/Display.html?id=83208
  

0.15 2015-08-11 Maintenance release, update not required

- Ported from Module::Build to Dist::Zilla [#1]


0.14 2015-08-09 Significant maintenance release, update recommended

- Addressed broken test suite [RT:94847]
  
  introducing the following method lifted from Module::Build::Base:

  - do_create_metafile

  https://rt.cpan.org/Ticket/Display.html?id=94847

- Core implementation updated to comply with Module::Build 0.42

- Changed license to be Artistic 2.0, please see the LICENSE file included
  with the distribution

- Examined and addressed issues with Module::Build compability, we are now
  dependent on version 0.42. New code lifted from Module::Build since
  proper inheritance is not possible at this time, pondering patches

- The whole test suite has been revisited and all tests now pass again
  as we rely on File::Tempdir instead of own logic, this eliminated a lot
  of code and removed at least one bug

- Addressed [Compatibility::PerlMinimumVersionAndWhy] _pragma_utf8
  requires 5.008, so perl requirement has been bumped up to version
  5.006.

- Added changes.t automatic assertion of the integrity of the Changes file
  see also MBB-38 [MBB-40]

  https://logiclab.jira.com/browse/MBB-40

- 0.13 release is a ghost it never got released

0.12 2013-08-01 Maintenance release, update not required

- Addressed issue reported by Ken Williams [RT:86220] so we now use CPAN::Meta::YAML
  instead of Module::Build::YAML [MBB-39]

  https://logiclab.jira.com/browse/MBB-39

- Fixed up Changes file as part of my Questhub quest adhering to the standard
  described in: CPAN::Changes::Spec [MBB-38]

  Quest: http://questhub.io/realm/perl/quest/51f0fd8e18ba7dcf4b000040
  Spec: https://metacpan.org/module/CPAN::Changes::Spec

  https://logiclab.jira.com/browse/MBB-38


0.11 2013-03-15 Maintenance release, update not required

- Addressed installation issue [MBB-37/[RT:83754] the issue was related to
  permissions on main file in lib. We now work on the equivalent in blib/lib.
  Issue was reported by DAXIM
  https://logiclab.jira.com/browse/MBB-37
  https://rt.cpan.org/Ticket/Display.html?id=83754


0.10 2013-01-26 Maintenance release, update not required

- Addressed failing tests for 0.09 [MBB-36]/[RT:82128], issue with undeclared
  dependency on File::Slurp, introduced in release 0.09. Issue found by CPAN
  testers and was also reported by ANDK
  https://logiclab.jira.com/browse/MBB-36
  https://rt.cpan.org/Ticket/Display.html?id=82128


0.09 2012-12-02 Maintenance release, update not required

- Attempt at address all the failing tests [MBB-35]
  http://logiclab.jira.com/browse/MBB-35

- Added Test::Perl::Critic to test suite

- Documentation on method ACTION_build [MBB-32]
  http://logiclab.jira.com/browse/MBB-32

- Replaced file reading code with use of File::Slurp as recommended by the
  Perl::Critic policy InputOutput::ProhibitJoinedReadline [MBB-34]
  http://logiclab.jira.com/browse/MBB-34


0.08 2012-11-22 Maintenance release, update not required

- Attempt at address all the failing tests (308 fails against 177 passes at
  the time of writing) [MBB-31]
  http://logiclab.jira.com/browse/MBB-31

- Turned the dates in the changes file into ISO 8601 format


0.07 2011-06-15 Maintenance release, update not required

- Addressed issue with test suite and permissions [MBB-28]
  http://logiclab.jira.com/browse/MBB-28

- Investigated create_makefile_pl [MBB-29]
  http://logiclab.jira.com/browse/MBB-29

- Described current limitation in POD in regard to root based distributions	
  [MBB-30]
  http://logiclab.jira.com/browse/MBB-30


0.06 2011-05-06 Maintenance release, update not required

- Addressed failing tests [MBB-25] tests should now pass under newer perls
  [MBB-26], when we cheat on the Perl version
  http://logiclab.jira.com/browse/MBB-25 and
  http://logiclab.jira.com/browse/MBB-26


0.05 2011-04-23 Bug fix release, update not required

- Addressed failing tests [MBB-22] removing regression test suite, M::B is
  a  moving target
  http://logiclab.jira.com/browse/MBB-22


0.04 2010-11-23 Bug fix release, update recommended

- Addressed failing tests [MBB-18] missing listing of required file
  http://logiclab.jira.com/browse/MBB-18


0.03 2010-11-21 Bug fix release, update recommended

- Addressed failing tests [MBB-17]
  http://logiclab.jira.com/browse/MBB-17


0.02 2010-09-26 Bug fix release, update recommended

- Changed inheritance [MBB-13]
  http://logiclab.jira.com/browse/MBB-13

- Issues with distribution generation [MBB-16]
  http://logiclab.jira.com/browse/MBB-16

- Addressed failing tests [MBB-14]
  http://logiclab.jira.com/browse/MBB-14


0.01 2010-09-14

- Initial release


t/Dummy.pm
package Dummy;

our $VERSION = '0.01';

1;

__END__

=head1 CONTENTS

=head1 ANOTHER SECTION

=cut


Build.PL
# Build.PL for Module::Build::Bundle

use strict;
use warnings;
use Module::Build;

our $VERSION = '0.02';

my $build = Module::Build->new(
    dist_author   => 'Jonas B. Nielsen (jonasbn), <jonasbn@cpan.org>',
    module_name   => 'Module::Build::Bundle',
    dist_abstract => 'build assistance for Bundle and Task distributions',
    license       => 'artistic_2',
    create_makefile_pl => 'traditional',
    create_readme      => 1,
    configure_requires => { 'Module::Build' => '0.42' },
    test_requires      => {
        'Env'                       => '0',
        'File::Copy'                => '0',
        'File::Slurp'               => '0',
        'File::Stat'                => '0',
        'File::Tempdir'             => '0',
        'FindBin'                   => '0',
        'CPAN::Meta::YAML'          => '0',
        'Test::Class'               => '0',
        'Test::More'                => '0',
        'Test::Exception'           => '0',
        'File::Tempdir'             => '0',
        'Test::MockObject::Extends' => '0',
    },
    requires => {
        'Carp'          => '0',
        'Cwd'           => '0',
        'English'       => '0',
        'File::Slurp'   => '0',
        'Tie::IxHash'   => '0',
        'perl'          => '5.8.0',
        'Module::Build' => '0.42',
        'Env'                       => '0',
        'File::Copy'                => '0',
        'File::Tempdir'             => '0',
        'FindBin'                   => '0',
        'CPAN::Meta::YAML'          => '0',
        'Test::Class'               => '0',
        'Test::More'                => '0',
        'Test::Exception'           => '0',
        'File::Tempdir'             => '0',
        'Test::MockObject::Extends' => '0',
    },
    meta_merge => {
        resources => {
            homepage => 'https://github.com/jonasbn/mbb',
            bugtracker =>
                'http://rt.cpan.org/NoAuth/Bugs.html?Dist=Module-Build-Bundle',
            repository => 'https://github.com/jonasbn/mbb',
        }
    },
);

$build->create_build_script();
